interface/src/ui/AssetUploadDialogFactory.cpp

```diff
@@ -0,0 +1,154 @@
+//
+//  AssetUploadDialogFactory.cpp
+//  interface/src/ui
+//
+//  Created by Stephen Birarda on 2015-08-26.
+//  Copyright 2015 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#include "AssetUploadDialogFactory.h"
+
+#include <AssetClient.h>
+#include <AssetUpload.h>
+#include <AssetUtils.h>
+#include <NodeList.h>
+
+#include <QtCore/QDebug>
+#include <QtWidgets/QDialogButtonBox>
+#include <QtWidgets/QFileDialog>
+#include <QtWidgets/QLabel>
+#include <QtWidgets/QMessageBox>
+#include <QtWidgets/QLineEdit>
+#include <QtWidgets/QVBoxLayout>
+
+AssetUploadDialogFactory& AssetUploadDialogFactory::getInstance() {
+    static AssetUploadDialogFactory staticInstance;
+    return staticInstance;
+}
+
+AssetUploadDialogFactory::AssetUploadDialogFactory() {
+    
+}
+
+static const QString PERMISSION_DENIED_ERROR = "You do not have permission to upload content to this asset-server.";
+
+void AssetUploadDialogFactory::showDialog() {
+    auto nodeList = DependencyManager::get<NodeList>();
+    
+    if (nodeList->getThisNodeCanRez()) {
+        auto filename = QFileDialog::getOpenFileUrl(_dialogParent, "Select a file to upload");
+        
+        if (!filename.isEmpty()) {
+            qDebug() << "Selected filename for upload to asset-server: " << filename;
+            
+            auto assetClient = DependencyManager::get<AssetClient>();
+            auto upload = assetClient->createUpload(filename.path());
+            
+            if (upload) {
+                // connect to the finished signal so we know when the AssetUpload is done
+                QObject::connect(upload, &AssetUpload::finished, this, &AssetUploadDialogFactory::handleUploadFinished);
+                
+                // start the upload now
+                upload->start();
+            } else {
+                // show a QMessageBox to say that there is no local asset server
+                QString messageBoxText = QString("Could not upload \n\n%1\n\nbecause you are currently not connected" \
+                                                 " to a local asset-server.").arg(QFileInfo(filename.toString()).fileName());
+                
+                QMessageBox::information(_dialogParent, "Failed to Upload", messageBoxText);
+            }
+        }
+    } else {
+        // we don't have permission to upload to asset server in this domain - show the permission denied error
+        showErrorDialog(QString(), PERMISSION_DENIED_ERROR);
+    }
+    
+}
+
+void AssetUploadDialogFactory::handleUploadFinished(AssetUpload* upload, const QString& hash) {
```

Comment:
Do we delete the upload object anywhere in here? I think we should call `upload->deleteLater()` at the end